src/Tribe/Utils/Collection_Trait.php

```diff
@@ -152,14 +152,27 @@ public function count() {
 	 * {@inheritDoc}
 	 */
 	public function serialize() {
-		return serialize( $this->all() );
+		$to_serialize = $this->all();
+
+		if ( method_exists( $this, 'before_serialize' ) ) {
+			$to_serialize = $this->before_serialize( $this->all() );
+		}
+
+		return serialize( $to_serialize );
```

Comment:
(2/2) If we _do_ need to perform wakeups and have to support direct serialization/unserialization of arbitrary classes etc this won't be tenable but, following my earlier question, might an alternative serialization technique (to JSON, for example) be worthwhile on the basis of improved security (noting that the mitigation given in the first part of this question won't fly on PHP 5.6).

Reply:
The problem with JSON is that the information about the serialized class is lost.

This is the result of a call to `json_encode` on a post:

```json
{"ID":381,"post_author":"1","post_date":"2019-12-17 08:51:58","post_date_gmt":"2019-12-17 07:51:58","post_content":"Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.","post_title":"CoworkingRepublic","post_excerpt":"","post_status":"publish","comment_status":"closed","ping_status":"closed","post_password":"","post_name":"coworkingrepublic","to_ping":"","pinged":"","post_modified":"2019-12-17 08:51:58","post_modified_gmt":"2019-12-17 07:51:58","post_content_filtered":"","post_parent":0,"guid":"http:\/\/mtvr.local\/?post_type=tribe_venue&#038;p=381","menu_order":0,"post_type":"tribe_venue","post_mime_type":"","comment_count":"0","filter":"raw"}
```

Furthermore, we attach objects to post properties, with `serialize` and `unserialize` we either get those implementations for free, or we can have them handled *if* we need to.

I understand the security concern here, yet we're reading from cached values, not user inputs.
I fully back the "trust no-one" policy, but I think we're pretty safe here.

Reply:
Losing the class information at serialization time means we, now, would have to pass that information in the class `__construct` method.
Which would make the use of the generic collection not so convenient.